File: dcmgadu/VIETBAIT.DICOMPDU/PDVItem.vb
Imports System.Text
Imports VIETBAIT.DICOM.BASE


Public Class PDVItem
    Dim ItemLen As UInt32
    Dim PreCtxID As Byte
    Dim MessageCtlHeader As Byte
    Public DataElements As New ArrayList()
    'Public DataElementHashTable As Hashtable

    Public Property Length() As UInt32
        Get
            Return ItemLen
        End Get
        Set(ByVal value As UInt32)
            ItemLen = value
        End Set
    End Property

    Public Property PresentationContextID() As Byte
        Get
            Return PreCtxID
        End Get
        Set(ByVal value As Byte)
            PreCtxID = value
        End Set
    End Property

    Property MessageControlHeader() As Byte
        Get
            Return MessageCtlHeader
        End Get
        Set(ByVal value As Byte)
            MessageCtlHeader = value
        End Set
    End Property

    Function IsCommand() As Boolean

        If (MessageControlHeader And &H1) Then
            Return True
        Else
            Return False
        End If
    End Function

    Function IsLast() As Boolean

        If (MessageControlHeader And &H2) Then
            Return True
        Else
            Return False
        End If
    End Function

    Sub Command(ByVal IsCom As Boolean)
        If IsCom Then
            MessageCtlHeader = MessageCtlHeader Or &H1
        Else
            MessageCtlHeader = MessageCtlHeader And &HFE
        End If
    End Sub

    Sub Last(ByVal IsLast As Boolean)
        If IsLast Then
            MessageCtlHeader = MessageCtlHeader Or &H2
        Else
            MessageCtlHeader = MessageCtlHeader And &HFD
        End If
    End Sub

    Function GetLength(ByVal b As Byte(), ByVal off As UInt32) As UInt32
        Dim tmp(3) As Byte
        Array.Copy(b, off, tmp, 0, 4)
        Array.Reverse(tmp)
        Return (BitConverter.ToUInt32(tmp, 0))
    End Function

    'Public Overloads Sub AddDataSet(ByVal sData As String, ByVal TagGroup As UInt16, ByVal TagElement As UInt16)
    '    Dim TagTmp As New DataElement
    '    TagTmp.DataTag = (TagGroup << 16) + TagElement

    '    If Not String.IsNullOrEmpty(sData) Then
    '        If sData.Length Mod 2 = 1 Then
    '            sData = sData & " "
    '        End If
    '        TagTmp.DataValue = Encoding.ASCII.GetBytes(sData)
    '        TagTmp.Length = sData.Length
    '    Else
    '        TagTmp.Length = 0
    '    End If
    '    DataElements.Add(TagTmp)
    'End Sub

    Public Overloads Sub AddDataSet(ByVal sData As String, ByVal DataTag As UInt32)
        Dim TagTmp As New DataElement
        TagTmp.DataTag = DataTag

        If Not String.IsNullOrEmpty(sData) Then
            If sData.Length Mod 2 = 1 Then
                sData = sData & " "
            End If
            TagTmp.DataValue = Encoding.ASCII.GetBytes(sData)
            TagTmp.Length = sData.Length
        Else
            TagTmp.Length = 0
        End If
        DataElements.Add(TagTmp)

    End Sub

    'Overloads Sub AddDataSet(ByVal iData As UInt16, ByVal TagGroup As UInt16, ByVal TagElement As UInt16)
    '    Dim TagTmp As New DataElement
    '    TagTmp.TagGroup = TagGroup
    '    TagTmp.TagElement = TagElement
    '    TagTmp.DataValue = BitConverter.GetBytes(CUShort(iData))
    '    TagTmp.Length = 2
    '    DataElements.Add(TagTmp)
    'End Sub

    'Overloads Sub AddDataSet(ByVal iData As UInt32, ByVal TagGroup As UInt16, ByVal TagElement As UInt16)
    '    Dim TagTmp As New DataElement
    '    TagTmp.TagGroup = TagGroup
    '    TagTmp.TagElement = TagElement
    '    TagTmp.DataValue = BitConverter.GetBytes(CUInt(iData))
    '    TagTmp.Length = 4
    '    DataElements.Add(TagTmp)
    'End Sub

    Overloads Sub AddDataSet(ByVal iData As UInt32, ByVal DataTag As UInt32)
        Dim TagTmp As New DataElement
        TagTmp.DataTag = DataTag
        TagTmp.DataValue = BitConverter.GetBytes(CUInt(iData))
        TagTmp.Length = 4
        DataElements.Add(TagTmp)
    End Sub

    Overloads Sub AddDataSet(ByVal iData As UInt16, ByVal DataTag As UInt32)
        Dim TagTmp As New DataElement
        TagTmp.DataTag = DataTag
        TagTmp.DataValue = BitConverter.GetBytes(CUShort(iData))
        TagTmp.Length = 2
        DataElements.Add(TagTmp)
    End Sub

    Sub AddDelimiterDataSet(ByVal _Tag As UInt32)

        Dim TagTmp As New DataElement

        TagTmp.DataTag = _Tag        
        TagTmp.Length = &HFFFFFFFFUI
        DataElements.Add(TagTmp)
    End Sub

    Sub AddZeroLenDataSet(ByVal _Tag As UInt32)
        Dim TagTmp As New DataElement
        TagTmp.DataTag = _Tag
        TagTmp.Length = &H0UI
        DataElements.Add(TagTmp)
    End Sub

    Sub CreateGroupLengthDataset(ByVal GroupLengthTag As UInt32)
        Dim de As New DataElement
        Dim GrLen As UInt32 = 0
        For Each de In DataElements
            If de.Length <> &HFFFFFFFFUI Then
                GrLen += de.Length + 8
            Else
                GrLen += 12
            End If
        Next
        de = New DataElement
        de.DataTag = GroupLengthTag        
        de.Length = 4
        de.DataValue = BitConverter.GetBytes(GrLen)
        DataElements.Insert(0, de)
    End Sub


    Function Parse(ByVal ByteBuff As Byte(), ByVal off As UInt16) As Byte()
        Dim pos As UInt32
        PreCtxID = ByteBuff(off)
        MessageCtlHeader = ByteBuff(off + 1)

        pos = off + 2
        While (pos < Length + off)
            If pos > Length + off - 8 Then
                Exit While
            End If
            Dim de As New DataElement

            de.DataTag = (BitConverter.ToUInt16(ByteBuff, pos) << 16) + BitConverter.ToUInt16(ByteBuff, pos + 2)            
            de.Length = BitConverter.ToUInt32(ByteBuff, pos + 4)
            Debug.WriteLine(Format(de.DataTag, "X8"))

            If (de.Length > 0) And (de.Length <> &HFFFFFFFFUI) Then
                If pos + de.Length + 8 > Length + off Then
                    Exit While
                End If
                ReDim de.DataValue(de.Length - 1)
                Array.Copy(ByteBuff, pos + 8, de.DataValue, 0, de.Length)
            Else
                de.Length = 0
            End If
            pos += de.Length + 8
            If (de.Length < 4) Then
                DataElements.Add(de)
            ElseIf (BitConverter.ToUInt32(de.DataValue, 0) <> &HE000FFFEUL) Then
                DataElements.Add(de)
            Else
                Dim pos1 As UInt32 = 8
                Dim len1 As UInt32 = BitConverter.ToUInt32(de.DataValue, 4)
                While (pos1 < len1 + 8)
                    Dim de1 As New DataElement
                    If BitConverter.ToUInt32(de.DataValue, pos1) = &HE00DFFFEUL Then
                        Exit While
                    End If
                    de1.DataTag = (BitConverter.ToUInt16(ByteBuff, pos) << 16) + BitConverter.ToUInt16(ByteBuff, pos + 2)                    
                    de1.Length = BitConverter.ToUInt32(de.DataValue, pos1 + 4)
                    Debug.WriteLine(Format(de1.DataTag, "X8"))
                    If de1.Length > 0 Then
                        ReDim de1.DataValue(de1.Length - 1)
                        Array.Copy(de.DataValue, pos1 + 8, de1.DataValue, 0, de1.Length)
                    End If
                    DataElements.Add(de1)
                    pos1 += de1.Length + 8
                End While
            End If

        End While
        If pos = Length + off Then
            Return Nothing
        Else
            Dim tmp(Length + off - pos - 1) As Byte
            Array.Copy(ByteBuff, pos, tmp, 0, Length + off - pos)
            Return tmp
        End If
    End Function

    
    Function CreateDataElementsBuffer() As Byte()
        Dim BB(4) As Byte
        Dim pos As UInt32 = 0
        Dim len As UInt32 = 0
        For Each de As DataElement In DataElements
            Debug.WriteLine(Format(de.DataTag, "X8"))
            Dim tmp1() As Byte = de.CreateByteBuff()
            len += de.Length + 8
            Array.Resize(BB, len)
            Array.Copy(tmp1, 0, BB, pos, tmp1.Length)
            pos += tmp1.Length
        Next
        Return BB
    End Function
    'Function CreateHashTableBuffer() As Byte()
    '    Dim BB(4) As Byte
    '    Dim pos As UInt32 = 0
    '    Dim len As UInt32 = 0
    '    Dim tagList As New ArrayList(hashDataElement.Keys)

    '    For Each _tag As UInt32 In tagList
    '        Debug.WriteLine(Format(_tag, "X8"))
    '        Dim de As New DataElement
    '        de.DataTag = _tag
    '        de.DataValue = hashDataElement(_tag)
    '        de.Length = de.DataValue.Length
    '        Dim tmp1() As Byte = de.CreateByteBuff()
    '        len += de.Length + 8
    '        Array.Resize(BB, len)
    '        Array.Copy(tmp1, 0, BB, pos, tmp1.Length)
    '        pos += tmp1.Length
    '    Next
    '    Return BB
    'End Function
    'Function CreateBufferFromHashtable() As Byte()
    '    Return CreateByteBuffer(CreateHashTableBuffer, 4)
    'End Function

    Overloads Function CreateByteBuffer() As Byte()
        Return CreateByteBuffer(CreateDataElementsBuffer, 4)
    End Function

    Overloads Function CreateByteBuffer(ByVal buffer As Byte(), ByVal off As UInt32) As Byte()
        Dim BB(off + 2 + buffer.Length - 1) As Byte

        BB(off) = PreCtxID
        BB(off + 1) = MessageCtlHeader
        Array.Copy(buffer, 0, BB, off + 2, buffer.Length)
        Length = buffer.Length + 2
        Dim tmp() As Byte = BitConverter.GetBytes(MyClass.Length)
        Array.Reverse(tmp)
        Array.Copy(tmp, BB, 4)
        Return BB
    End Function

    'Overloads Function DataTagSearch(ByVal DataTagKey As UInt16) As Int32
    '    Dim Index As UInt32
    '    Dim de As DataElement
    '    For Index = 0 To DataElements.Count - 1
    '        de = DataElements(Index)
    '        If de.TagElement = DataTagKey Then
    '            Return Index
    '        End If
    '    Next
    '    Return -1
    'End Function

    Overloads Function DataTagSearch(ByVal DataTagKey As UInt32) As Int32
        Dim Index As UInt32
        Dim de As DataElement
        For Index = 0 To DataElements.Count - 1
            de = DataElements(Index)
            If de.DataTag = DataTagKey Then
                Return Index
            End If
        Next
        Return -1
    End Function
End Class

File: dcmgadu/VIETBAIT.DICOM.BASE/PDVItem.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Diagnostics;
using System.Text;

namespace VIETBAIT.DICOM.BASE
{
    public class PDVItem
    {
        public List<DataElement> DataElements {get;set;} 
        public Hashtable DataElementHashTable{get;set;}

        public UInt32 Length { get; set; }


        public byte PresentationContextID { get; set; }


        public byte MessageControlHeader { get; set; }

        public PDVItem ()
        {
            DataElements= new List<DataElement>();
            DataElementHashTable = new Hashtable();
        }
        ~PDVItem ()
        {
            DataElements.Clear();
            DataElements = null;
            DataElementHashTable.Clear();
            DataElementHashTable = null;
        }

        public bool IsCommand()
        {
            if ((MessageControlHeader & 0x01) != 0)
            {
                return true;
            }
            else
            {
                return false;
            }
        }


        public bool IsLast()
        {
            if ((MessageControlHeader & 0x2) != 0)
            {
                return true;
            }
            else
            {
                return false;
            }
        }

        public void Command(bool isCom)
        {
            if (isCom)
            {
                MessageControlHeader |= (0x01);
            }
            else
            {
                MessageControlHeader &= 0xFE;
            }
        }

        public void Last(bool isLast)
        {
            if (isLast)
            {
                MessageControlHeader |= 0x2;
            }
            else
            {
                MessageControlHeader &= 0xFD;
            }
        }

        public UInt32 GetLength(byte[] b, UInt32 off)
        {
            var tmp = new byte[4];
            Array.Copy(b, off, tmp, 0, 4);
            Array.Reverse(tmp);
            return (BitConverter.ToUInt32(tmp, 0));
        }

        //Public Overloads Sub AddDataSet(ByVal sData As String, ByVal TagGroup As UInt16, ByVal TagElement As UInt16)
        //    Dim TagTmp As New DataElement
        //    TagTmp.DataTag = (TagGroup << 16) + TagElement

        //    If Not String.IsNullOrEmpty(sData) Then
        //        If sData.Length Mod 2 = 1 Then
        //            sData = sData & " "
        //        End If
        //        TagTmp.DataValue = Encoding.ASCII.GetBytes(sData)
        //        TagTmp.Length = sData.Length
        //    Else
        //        TagTmp.Length = 0
        //    End If
        //    DataElements.Add(TagTmp)
        //End Sub

        public void AddDataSet(string sData, UInt32 dataTag)
        {
            var tagTmp = new DataElement();
            tagTmp.DataTag = dataTag;

            if (!string.IsNullOrEmpty(sData))
            {
                if (sData.Length%2 == 1)
                {
                    sData = sData + " ";
                }
                tagTmp.DataValue = Encoding.ASCII.GetBytes(sData);
                tagTmp.Length = (UInt32) sData.Length;
            }
            else
            {
                tagTmp.Length = 0;
            }
            DataElements.Add(tagTmp);
        }

        //Overloads Sub AddDataSet(ByVal iData As UInt16, ByVal TagGroup As UInt16, ByVal TagElement As UInt16)
        //    Dim TagTmp As New DataElement
        //    TagTmp.TagGroup = TagGroup
        //    TagTmp.TagElement = TagElement
        //    TagTmp.DataValue = BitConverter.GetBytes(CUShort(iData))
        //    TagTmp.Length = 2
        //    DataElements.Add(TagTmp)
        //End Sub

        //Overloads Sub AddDataSet(ByVal iData As UInt32, ByVal TagGroup As UInt16, ByVal TagElement As UInt16)
        //    Dim TagTmp As New DataElement
        //    TagTmp.TagGroup = TagGroup
        //    TagTmp.TagElement = TagElement
        //    TagTmp.DataValue = BitConverter.GetBytes(CUInt(iData))
        //    TagTmp.Length = 4
        //    DataElements.Add(TagTmp)
        //End Sub

        public void AddDataSet(UInt32 iData, UInt32 dataTag)
        {
            var tagTmp = new DataElement();
            tagTmp.DataTag = dataTag;
            tagTmp.DataValue = BitConverter.GetBytes(Convert.ToUInt32(iData));
            tagTmp.Length = 4;
            DataElements.Add(tagTmp);
        }

        public void AddDataSet(UInt16 iData, UInt32 dataTag)
        {
            var tagTmp = new DataElement();
            tagTmp.DataTag = dataTag;
            tagTmp.DataValue = BitConverter.GetBytes(Convert.ToUInt16(iData));
            tagTmp.Length = 2;
            DataElements.Add(tagTmp);
        }


        public void AddDelimiterDataSet(UInt32 tag)
        {
            var tagTmp = new DataElement(tag,(UInt32 ) 0xFFFFFFFFU);            
            DataElements.Add(tagTmp);
        }

        public void AddZeroLenDataSet(UInt32 tag)
        {
            var tagTmp = new DataElement(tag,(UInt32)0);            
            DataElements.Add(tagTmp);
        }

        public void CreateGroupLengthDataset(UInt32 groupLengthTag)
        {
            UInt32 grLen = 0;
            foreach (DataElement dataElement  in DataElements)
            {
                if (dataElement.Length != 0xFFFFFFFFU)
                {
                    grLen += dataElement.Length + 8;
                }
                else
                {
                    grLen += 12;
                }
            }
            var de = new DataElement();
            de.DataTag = groupLengthTag;
            de.Length = 4;
            de.DataValue = BitConverter.GetBytes(grLen);
            DataElements.Insert(0, de);
        }


        public  byte[] Parse(byte[] byteBuff, UInt16 off)
        {
            if (byteBuff == null) throw new ArgumentNullException("byteBuff");
            try
            {
                
                UInt32 pos;
                PresentationContextID = byteBuff[off];
                MessageControlHeader = byteBuff[off + 1];

                pos = (UInt32)off + 2;
                while ((pos < Length))
                {
                    if (pos > Length + off - 8)
                    {
                        break;
                    }


                    UInt32 key = (UInt32)((BitConverter.ToUInt16(byteBuff, (int)pos)) << 16) +
                                 BitConverter.ToUInt16(byteBuff, (int)pos + 2);
                    UInt32 Len = BitConverter.ToUInt32(byteBuff, (int)pos + 4);
                    var de = new DataElement();
                    Debug.WriteLine(key.ToString("X8"));
                    if (key.Equals(0xFFFEE0DDU) || key.Equals(0xFFFEE00DU) || key.Equals(0xFFFEE000U) ||
                        (Len == 0xFFFFFFFFU))
                    {
                        pos += 8;
                        //if (Len == 0xFFFFFFFFU) DataElementHashTable.Add(key, null);
                        de = new DataElement(key, 0);
                        DataElements.Add(de);
                        continue;
                    }
                    if (Len > 0)
                    {
                        if (pos + Len + 8 > Length + off)
                        {
                            break;
                        }

                        var dataValue = new byte[Len];
                        Array.Copy(byteBuff, pos + 8, dataValue, 0, Len);
                        if (Len > 7)
                        {
                            if (BitConverter.ToUInt32(dataValue,0) == 0xe000fffeuL)
                            {
                                Len = 0;
                                DataElementHashTable.Add(key, null );
                                de = new DataElement(key, Len );
                                DataElements.Add(de);
                            }
                            else
                            {
                                DataElementHashTable.Add(key, dataValue);
                                de = new DataElement(key, dataValue);
                                DataElements.Add(de);
                            }
                        }
                        else
                        {
                            DataElementHashTable.Add(key, dataValue);
                            de = new DataElement(key, dataValue);
                            DataElements.Add(de);
                        }
                    }
                    else
                    {
                        Len = 0;
                        DataElementHashTable.Add(key, null);
                        de = new DataElement(key, 0);
                        DataElements.Add(de);
                    }
                    pos += Len + 8;
                }

                if (pos == Length + off)
                {
                    return null;
                }
                else
                {
                    var tmp = new byte[Length + off - pos];
                    Array.Copy(byteBuff, pos, tmp, 0, Length + off - pos);
                    return tmp;
                }
            }
            catch (Exception exception)
            {

                throw exception;
            }
        }

        


        private byte[] CreateDataElementsBuffer()
        {
            var buffer = new byte[4];
            UInt32 pos = 0;
            UInt32 len = 0;
            foreach (DataElement de in DataElements)
            {
                Debug .WriteLine(de.DataTag.ToString("X8"));

                byte[] tmp1 = de.CreateByteBuff();
                len += (UInt32)tmp1.Length;
                Array.Resize<byte>(ref buffer, (int) len);              
                Array.Copy(tmp1, 0, buffer, pos, tmp1.Length);
                pos += (UInt32) tmp1.Length;
            }
            return buffer;
        }

        //Function CreateHashTableBuffer() As Byte()
        //    Dim BB(4) As Byte
        //    Dim pos As UInt32 = 0
        //    Dim len As UInt32 = 0
        //    Dim tagList As New ArrayList(hashDataElement.Keys)

        //    For Each _tag As UInt32 In tagList
        //        Debug.WriteLine(Format(_tag, "X8"))
        //        Dim de As New DataElement
        //        de.DataTag = _tag
        //        de.DataValue = hashDataElement(_tag)
        //        de.Length = de.DataValue.Length
        //        Dim tmp1() As Byte = de.CreateByteBuff()
        //        len += de.Length + 8
        //        Array.Resize(BB, len)
        //        Array.Copy(tmp1, 0, BB, pos, tmp1.Length)
        //        pos += tmp1.Length
        //    Next
        //    Return BB
        //End Function
        //Function CreateBufferFromHashtable() As Byte()
        //    Return CreateByteBuffer(CreateHashTableBuffer, 4)
        //End Function
        public byte[] CreateByteBuffer(byte[] buffer, UInt32 off)
        {
            var bb = new byte[off + 2 + buffer.Length];

            bb[off] = PresentationContextID;
            bb[off + 1] = MessageControlHeader;
            Array.Copy(buffer, 0, bb, off + 2, buffer.Length);
            Length = (UInt32)buffer.Length + 2;
            byte[] tmp = BitConverter.GetBytes(Length);
            Array.Reverse(tmp);
            Array.Copy(tmp, bb, 4);
            return bb;
        }

        public byte[] CreateByteBuffer()
        {
            return CreateByteBuffer(CreateDataElementsBuffer(), 4);
        }

        

        //Overloads Function DataTagSearch(ByVal DataTagKey As UInt16) As Int32
        //    Dim Index As UInt32
        //    Dim de As DataElement
        //    For Index = 0 To DataElements.Count - 1
        //        de = DataElements(Index)
        //        If de.TagElement = DataTagKey Then
        //            Return Index
        //        End If
        //    Next
        //    Return -1
        //End Function

        public DataElement DataTagSearch(UInt32 dataTagKey)
        {
            DataElement de = DataElements.Find(
                dataElement => dataElement.DataTag == dataTagKey
                );
            return de;
        }
        
    }
}